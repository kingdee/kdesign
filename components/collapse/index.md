---
category: Components
type: 数据展示
order: 3
title: Collapse
subtitle: 折叠面板
---

折叠面板是一种对内容进行折叠和展开的容器。

## 使用场景

需要对复杂的内容进行分组和隐藏时，保证页面的整洁。

## API

### Collapse

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| accordion | 是否手风琴模式 | boolean | false | true \| false | 1.0.0 |
| activeKey | 当前激活 tab 面板的 key | string[] \| string \| number[] \| number | - | - | 1.0.0 |
| bordered | 是否边框风格折叠面板是否边框风格折叠面板 | boolean | false | true \| false | 1.0.0 |
| defaultActiveKey | 初始化选中面板的 key | string[] \| string \| number[] \| number | - | - | 1.0.0 |
| expandIcon | 自定义切换图标 | (panelProps) => ReactNode | - | - | 1.0.0 |
| expandIconPosition | 设置切换图标位置 | string | `left` | `left` \| `right` | 1.0.0 |
| onChange           | 切换面板时的回调 | function | - | - | 1.0.0 |

### Collapse.Panel

| 属性     | 说明                           | 类型      | 默认值 | 可选值        | 版本   |
| -------- | ------------------------------ | --------- | ------ | ------------- | ------ |
| disabled | 禁用                           | boolean   | false  | true \| false | 1.0.0 |
| extra    | 自定义渲染每个面板右上角的内容 | ReactNode | -      | -             | 1.0.0 |
| header   | 折叠头内容                     | ReactNode | -      | -             | 1.0.0 |
| key      | 对应 Collapse 的 activeKey     | string \| number | -      | -             | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-collapse-color-border-strong-2 | --kd-g-color-border-strong-2 | #d9d9d9 |
|  | --kd-c-collapse-header-color-text-primary | --kd-g-color-text-primary | #212121 |
|  | --kd-c-collapse-header-color-text-secondary | --kd-g-color-text-secondary | #666 |
|  | --kd-c-collapse-content-color-text-primary | --kd-g-color-text-primary | #212121 |
|  | --kd-c-collapse-color-disabled | - | raba(178, 178, 178, 1) |
|  | --kd-c-collapse-header-color-background | - | rgba(0, 0, 0, 0) |
|  | --kd-c-collapse-content-color-background | - | rgba(0, 0, 0, 0) |
|  | --kd-c-collapse-header-color-text-hover | - | #666666 |
| font | --kd-c-collapse-header-font-size | - | 16px |
|  | --kd-c-collapse-content-font-size | --kd-g-font-size-middle | 14px |
|  | --kd-c-collapse-icon-font-size | - | 16px |
| motion | --kd-c-collapse-motion-duration | --kd-g-duration | 0.3s |
| sizing | --kd-c-collapse-border-width | - | 1px |
|  | --kd-c-collapse-spacing-margin-right | - | 6px |
|  | --kd-c-collapse-header-spacing-padding | - | 8px 50px 8px 28px |
|  | --kd-c-collapse-content-bordered-spacing-padding | - | 12px |
|  | --kd-c-collapse-content-spacing-padding | - | 0 50px |
