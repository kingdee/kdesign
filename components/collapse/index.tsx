import InternalCollapse, { CollapseProps, IconPositionType } from './collapse'
import Panel from './panel'
export { PanelProps } from './panel'
interface CompoundedComponent
  extends React.ForwardRefExoticComponent<CollapseProps & React.RefAttributes<HTMLElement>> {
  Panel: typeof Panel
}
const Collapse = InternalCollapse as CompoundedComponent
Collapse.Panel = Panel
export { Panel, IconPositionType }
export default Collapse
