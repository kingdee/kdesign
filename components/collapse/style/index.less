@import '../../style/themes/index';
@import './mixin.less';

@collapse-prefix-cls: ~'@{kd-prefix}-collapse';
.@{collapse-prefix-cls} {
  display: flex;
  flex-direction: column;
  width: 100%;
  &-panel:last-child {
    border-top: none;
  }
  &-panel-border&-panel:first-child {
    border-top: @collapse-border-width solid @collapse-border-color;
  }
  &-panel-border&-panel:not(:nth-child(1)) {
    border-top: none;
  }
  &-panel {
    display: inline-flex;
    flex-direction: column;
    &-border {
      border: @collapse-border-width solid @collapse-border-color;
    }

    &-top {
      display: inline-flex;
      padding: @collapse-header-padding;
      background: @collapse-header-color-background;
    }
    &-top-border {
      border-bottom: @collapse-border-width solid @collapse-border-color;
    }
    &-left {
      display: inline-flex;
      align-items: center;
      line-height: 1.5;
    }
    &-middle {
      display: inline-flex;
      align-items: center;
      flex: 1;
      margin-left: 22px;
      color: @collapse-header-font-color-secondary;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    &-right {
      margin-left: 40px;
      display: inline-flex;
      align-items: center;
      white-space: nowrap;
    }
    &-extra,
    &-header {
      display: inline-flex;
      align-items: center;
      color: @collapse-header-font-color;
      font-size: @collapse-header-font-size;
    }

    &-header {
      display: inline-block;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      &:hover {
        color: @collapse-header-color-text-hover;
      }
    }
    &-icon {
      align-items: center;
      display: inline-flex;
      margin-right: @collapse-icon-margin-right;
      font-size: @collapse-icon-font-size;
    }
    &-icon:hover,
    &-left:hover {
      cursor: pointer;
    }
    &-animation-expand {
      transform: rotate(0deg);
      transition: transform @collapse-transition-duration;
    }
    &-animation-collapse {
      transform: rotate(90deg);
      transition: transform @collapse-transition-duration;
    }
    &-children {
      overflow: hidden;
      opacity: 0;
      padding: @collapse-content-padding;
      color: @collapse-content-font-color;
      font-size: @collapse-content-font-size;
      transition: height @collapse-transition-duration, opacity @collapse-transition-duration;
      background: @collapse-content-color-background;
    }
    &-children-hide {
      overflow: hidden;
      opacity: 0;
      height: 0;
      padding: @collapse-content-padding;
      color: @collapse-content-font-color;
      font-size: @collapse-content-font-size;
      transition: height @collapse-transition-duration, opacity calc(@collapse-transition-duration / 2);
    }
    &-children-bordered {
      padding: @collapse-content-bordered-padding;
      overflow: hidden;
    }
    &-disabled {
      cursor: not-allowed !important;
      color: @collapse-disabled-color;
    }
  }
}
