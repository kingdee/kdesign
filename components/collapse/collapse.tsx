import React, { useEffect, useState } from 'react'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import classNames from 'classnames'
import { tuple } from '../_utils/type'
import { PanelProps } from './panel'
import devwarning from '../_utils/devwarning'
export const IconPositionTypes = tuple('left', 'right')
export type IconPositionType = typeof IconPositionTypes[number]
export type PanelKeyType = string | number
export type keyType = string[] | string | number[] | number | undefined
export interface CollapseProps {
  accordion?: boolean // 是否手风琴模式
  activeKey?: keyType // 当前激活 tab 面板的 key
  bordered?: boolean // 是否边框风格折叠面板
  defaultActiveKey?: string[] | string | number[] | number // 初始化选中面板的 key
  expandIcon?: React.ReactNode | ((props: PanelProps) => React.ReactNode) // 自定义切换图标
  expandIconPosition?: IconPositionType // 设置切换图标位置
  onChange?: (v: any) => void // 切换面板时的回调
  style?: React.CSSProperties
  className?: string
  children?: React.ReactNode
}
const InternalCollapse = React.forwardRef<unknown, CollapseProps>((props, ref) => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = React.useContext(ConfigContext)
  const {
    accordion,
    activeKey,
    bordered,
    defaultActiveKey,
    expandIcon,
    expandIconPosition,
    onChange,
    style,
    className,
    children,
    prefixCls: customPrefixcls,
  } = getCompProps('Collapse', userDefaultProps, props)
  const CollapsePrefixCls = getPrefixCls!(prefixCls, 'collapse', customPrefixcls) // 样式前缀

  const [innerKey, setInnerKey] = useState<PanelKeyType[]>([])
  const convertActiveKey = (newKey: keyType) => {
    let ret: PanelKeyType[] = []
    if (Array.isArray(newKey)) {
      ret = newKey
    } else if (typeof newKey === 'string' || typeof newKey === 'number') {
      ret = [newKey]
    }
    if (accordion && ret.length) {
      ret = [ret[0]]
    }
    return ret
  }
  useEffect(() => {
    if (typeof activeKey !== 'undefined') {
      setInnerKey(convertActiveKey(activeKey))
    }
  }, [activeKey])
  useEffect(() => {
    if (typeof activeKey !== 'undefined' || typeof defaultActiveKey !== 'undefined') {
      setInnerKey(convertActiveKey(typeof activeKey !== 'undefined' ? activeKey : defaultActiveKey))
    }
  }, [])

  const onPanelChange = (key: PanelKeyType) => {
    let newKey: PanelKeyType[] = []
    if (accordion) {
      if (innerKey.includes(key)) {
        newKey = []
      } else {
        newKey = [key]
      }
    } else {
      if (innerKey.includes(key)) {
        newKey = innerKey.filter((d) => d !== key)
      } else {
        newKey = [...innerKey, key]
      }
    }

    if (typeof activeKey === 'undefined') {
      setInnerKey(newKey)
    }
    onChange && onChange(newKey)
  }

  const renderPanel = () => {
    return React.Children.map(children, (item: any) => {
      if (item?.type?.displayName !== 'Panel') {
        devwarning(true, 'Collapse', 'children必须为Collapse.Panel')
        return item
      }

      return React.cloneElement(item, {
        expandIcon,
        expandIconPosition,
        onChange: onPanelChange,
        bordered,
        innerKey,
        ...item.props,
      })
    })
  }

  const rootClassName = classNames(className, {
    [`${CollapsePrefixCls}`]: true,
  })

  const collapseRef = (ref as any) || React.createRef<HTMLElement>()

  return (
    <div className={rootClassName} style={style} ref={collapseRef}>
      {renderPanel()}
    </div>
  )
})

InternalCollapse.displayName = 'Collapse'
export default InternalCollapse
