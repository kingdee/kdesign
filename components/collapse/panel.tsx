import React, { memo, useEffect, useRef, useState } from 'react'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import Icon from '../icon'
import { IconPositionType, PanelKeyType } from './collapse'

export interface PanelProps {
  disabled?: boolean
  header?: React.ReactNode | ((props: PanelProps) => React.ReactNode)
  expandIcon?: React.ReactNode | ((props: PanelProps) => React.ReactNode) // 自定义切换图标
  bordered?: boolean // 是否边框风格折叠面板
  expandIconPosition?: IconPositionType // 设置切换图标位置
  onChange?: (key: PanelKeyType) => void // 切换面板时的回调
  extra?: React.ReactNode | ((props: PanelProps) => React.ReactNode)
  assist?: React.ReactNode | ((props: PanelProps) => React.ReactNode)
  expand?: boolean
  defaultExpand?: boolean
  children?: React.ReactNode
  panelKey?: PanelKeyType
  style?: React.CSSProperties
  className?: string
}
const InnerPanel = React.forwardRef<unknown, PanelProps>((props, ref) => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = React.useContext(ConfigContext)
  const {
    disabled,
    extra,
    assist,
    header,
    onChange,
    panelKey,
    isExpand,
    bordered,
    expandIcon,
    expandIconPosition,
    style,
    title,
    className,
    children,
    prefixCls: customPrefixcls,
    ...others
  } = getCompProps('CollapsePanel', userDefaultProps, props)

  const panelPrefixCls = getPrefixCls!(prefixCls, 'collapse-panel', customPrefixcls) // 样式前缀

  const childrenRef = useRef<HTMLElement>()
  const setHeightTimerRef = useRef<any>(null)
  const setAutoHeightTimerRef = useRef<any>(null)
  const [expandAnimation, setExpandAnimation] = useState<any>(null)
  const [expand, setExpand] = useState<boolean>(false)

  useEffect(() => {
    setExpand((expand) => {
      if (isExpand !== expand) return isExpand
    })
    setExpandAnimation(true)
  }, [isExpand])

  const handleClick = () => {
    if (disabled) return
    onChange && onChange(panelKey)
  }
  const renderIcon = () => {
    const iconClassName = classNames({
      [`${panelPrefixCls}-icon`]: true,
      [`${panelPrefixCls}-animation-expand`]: !expand,
      [`${panelPrefixCls}-animation-collapse`]: expand,
      [`${panelPrefixCls}-disabled`]: disabled,
    })
    return (
      <span className={iconClassName} onClick={handleClick}>
        {expandIcon ? renderReactNode(expandIcon) : <Icon type="arrow-right-solid" />}
      </span>
    )
  }

  const renderLeft = () => {
    const className = classNames({
      [`${panelPrefixCls}-left`]: true,
      [`${panelPrefixCls}-disabled`]: disabled,
    })
    const headerClassName = classNames({
      [`${panelPrefixCls}-header`]: true,
      [`${panelPrefixCls}-disabled`]: disabled,
    })
    return (
      <span className={className} onClick={handleClick}>
        {expandIconPosition === 'left' ? renderIcon() : null}
        {header && (
          <span className={headerClassName} title={title}>
            {renderReactNode(header)}
          </span>
        )}
      </span>
    )
  }

  const renderRight = () => {
    if (typeof extra === 'undefined' && expandIconPosition !== 'right') return
    const className = classNames({
      [`${panelPrefixCls}-right`]: true,
      [`${panelPrefixCls}-disabled`]: disabled,
    })
    return (
      <span className={className}>
        {extra && <span className={`${panelPrefixCls}-extra`}>{renderReactNode(extra)}</span>}
        {expandIconPosition === 'right' ? renderIcon() : null}
      </span>
    )
  }

  const renderMiddle = () => {
    if (typeof assist === 'undefined') return
    const className = classNames({
      [`${panelPrefixCls}-middle`]: true,
      [`${panelPrefixCls}-disabled`]: disabled,
    })
    return (
      <span className={className}>
        {assist && <span className={`${panelPrefixCls}-assist`}>{renderReactNode(assist)}</span>}
      </span>
    )
  }

  const renderReactNode = (reactNode: React.ReactNode) => {
    if (typeof reactNode === 'function') {
      return reactNode()
    }
    return reactNode
  }

  const rootClassName = classNames(className, {
    [`${panelPrefixCls}`]: true,
    [`${panelPrefixCls}-border`]: bordered,
    [`${panelPrefixCls}-opened`]: expand,
  })
  const topClassName = classNames({
    [`${panelPrefixCls}-top`]: true,
    [`${panelPrefixCls}-disabled`]: disabled,
    [`${panelPrefixCls}-top-border`]: bordered && expand,
    [`${panelPrefixCls}-no-assist`]: !assist,
  })
  const childrenClassName = classNames({
    [`${panelPrefixCls}-children`]: expand,
    [`${panelPrefixCls}-children-hide`]: !expand,
  })
  const childrenBorderedClassName = classNames({
    [`${panelPrefixCls}-children-bordered`]: bordered,
  })
  const runExpandAnimation = (element: HTMLElement) => {
    element.style.height = '0px'
    element.style.opacity = '0'
    setHeightTimerRef.current && clearTimeout(setHeightTimerRef.current)
    setHeightTimerRef.current = setTimeout(() => {
      element.style.display = 'block'
      element.style.opacity = '1'
      element.style.height = `${element.scrollHeight}px`
    }, 10)
    setAutoHeightTimerRef.current && clearTimeout(setAutoHeightTimerRef.current)
    setAutoHeightTimerRef.current = setTimeout(() => {
      element.style.height = 'auto'
    }, 350)
  }
  const runCollapseAnimation = (element: HTMLElement) => {
    setAutoHeightTimerRef.current && clearTimeout(setAutoHeightTimerRef.current)
    element.style.height = `${element.scrollHeight}px`
    setAutoHeightTimerRef.current = setTimeout(() => {
      element.style.height = '0px'
      element.style.opacity = '0'
    }, 0)
  }

  const expandWidthoutAnimation = (element: HTMLElement) => {
    element.style.height = 'auto'
    element.style.opacity = '1'
  }

  const panelPrefixClsRef = (ref as any) || React.createRef<HTMLElement>()

  React.useLayoutEffect(() => {
    if (!childrenRef.current) return
    if (expandAnimation) {
      expand ? runExpandAnimation(childrenRef.current) : runCollapseAnimation(childrenRef.current)
      setExpandAnimation(false)
    } else if (!expandAnimation) {
      expand ? expandWidthoutAnimation(childrenRef.current) : runCollapseAnimation(childrenRef.current)
    }
  }, [expandAnimation, expand])

  return (
    <div className={rootClassName} style={style} ref={panelPrefixClsRef} {...others}>
      <span className={topClassName}>
        {renderLeft()}
        {renderMiddle()}
        {renderRight()}
      </span>
      <div className={childrenClassName} ref={childrenRef as any}>
        <div className={childrenBorderedClassName}>{children}</div>
      </div>
    </div>
  )
})
const Panel = memo(InnerPanel)
Panel.displayName = 'Panel'
export default Panel
