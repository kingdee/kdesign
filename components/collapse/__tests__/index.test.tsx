import React from 'react'
import { mount, render } from 'enzyme'
import mountTest from '../../../tests/shared/mountTest'
import Collapse from '../index'
import Icon from '../../icon'
import { IconPositionTypes } from '../collapse'
import ConfigProvider from '../../config-provider/index'

const collapseProps = {
  accordion: false,
  bordered: false,
  onChange: jest.fn(),
}
const collapsePanelProps = {
  disabled: false,
}
describe('Image', () => {
  // 1. mount test
  describe('1. mount test', () => {
    mountTest(Collapse)
  })

  // 2. render test
  describe('2. render test', () => {
    it('renders correctly', () => {
      expect(
        render(
          <Collapse>
            <Collapse.Panel header={'This is the header'} panelKey="panel_1">
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        ),
      ).toMatchSnapshot()
    })
  })

  // 3. warns in component

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined without errors', () => {
      const wrapper = mount(
        <Collapse>
          <Collapse.Panel header={'This is the header'} panelKey="panel_1">
            {null}
            {undefined}
          </Collapse.Panel>
        </Collapse>,
      )
      // expect(wrapper).toBeEmptyRender()
      expect(wrapper.find('.kd-collapse')).toBeTruthy()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      const wrapper = mount(
        <Collapse>
          <Collapse.Panel header={'This is the header'} panelKey="panel_1">
            折叠面板内容
          </Collapse.Panel>
        </Collapse>,
      )
      expect((wrapper.type() as any).displayName).toBe('Collapse')
    })
  })

  // 6. api test
  describe('6. api test', () => {
    it('test icon position types', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header'}
              panelKey="panel_1"
            >
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        )
        expect(wrapper.find(`.kd-collapse-panel-${type}`).childAt(0)).toHaveClassName('.kd-collapse-panel-icon')
      })
    })

    it('test click event', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse {...collapseProps}>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header'}
              panelKey="panel_1"
            >
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        )
        wrapper.find(`.kd-collapse-panel-icon`).simulate('click')
        expect(collapseProps.onChange).toHaveBeenCalled()
      })
    })

    it('test accordion mode', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse {...collapseProps} accordion>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header1'}
              panelKey="panel_1"
            >
              折叠面板内容
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header2'}
              panelKey="panel_2"
            >
              折叠面板内容
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header3'}
              panelKey="panel_3"
            >
              折叠面板内容
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header4'}
              panelKey="panel_4"
            >
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        )
        wrapper.find(`.kd-collapse-panel-icon`).at(0).simulate('click')
        expect(wrapper.find(`.kd-collapse-panel-children`).length).toBe(1)
        wrapper.find(`.kd-collapse-panel-icon`).at(2).simulate('click')
        expect(wrapper.find(`.kd-collapse-panel-children`).length).toBe(1)
      })
    })

    it('test have border', () => {
      const wrapper = mount(
        <Collapse {...collapseProps} bordered={true} defaultActiveKey={['panel_1']}>
          <Collapse.Panel
            {...collapsePanelProps}
            expandIconPosition={'left'}
            header={'This is the header'}
            panelKey="panel_1"
          >
            折叠面板内容
          </Collapse.Panel>
        </Collapse>,
      )
      expect(wrapper.find(`.kd-collapse-panel-border`).length).toBe(1)
      expect(wrapper.find(`.kd-collapse-panel-top-border`).length).toBe(1)
    })

    it('customize expand icon', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse
            {...collapseProps}
            expandIcon={() => {
              return <Icon type="arrow-right"></Icon>
            }}
          >
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header'}
              panelKey="panel_1"
            >
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        )
        expect(wrapper.find(`.kdicon-arrow-right`).length).toBe(1)
      })
    })

    it('test default active key', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse {...collapseProps} defaultActiveKey="panel_1">
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header'}
              panelKey="panel_1"
              disabled
            >
              折叠面板内容
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header'}
              panelKey="panel_2"
              disabled
            >
              折叠面板内容
            </Collapse.Panel>
          </Collapse>,
        )
        expect(wrapper.find(`.kd-collapse-panel-children`).length).toBe(1)
      })
    })

    it('test disabled', () => {
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse
            {...collapseProps}
            defaultActiveKey="panel_1"
            expandIcon={() => {
              return <Icon type="arrow-right"></Icon>
            }}
          >
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header1'}
              panelKey="panel_1"
              disabled
            >
              折叠面板内容1
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header2'}
              panelKey="panel_2"
              disabled
            >
              折叠面板内容2
            </Collapse.Panel>
          </Collapse>,
        )
        expect(wrapper.find(`.kd-collapse-panel-disabled`).length).toBeGreaterThan(1)
        expect(wrapper.find(`.kd-collapse-panel-children`).length).toBe(1)
      })
    })

    it('customize upper right corner content', () => {
      const extra1 = (
        <div style={{ cursor: 'pointer', display: 'flex' }} className={'extra'}>
          <button>按钮</button>
          <p>|</p>
          <button>按钮</button>
          <p>|</p>
          <button>按钮</button>
          <p>|</p>
          <button>按钮</button>
        </div>
      )
      const extra2 = (
        <div>
          <Icon type="setting" />
        </div>
      )
      IconPositionTypes.forEach((type) => {
        const wrapper = mount(
          <Collapse {...collapseProps}>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header1'}
              panelKey="panel_1"
              extra={extra1}
            >
              折叠面板内容1
            </Collapse.Panel>
            <Collapse.Panel
              {...collapsePanelProps}
              expandIconPosition={type}
              header={'This is the header2'}
              panelKey="panel_2"
              extra={extra2}
            >
              折叠面板内容2
            </Collapse.Panel>
          </Collapse>,
        )
        expect(wrapper.find(`.extra`).length).toBe(1)
        expect(wrapper.find(`.kdicon-setting`).length).toBe(1)
      })
    })

    it('customize header', () => {
      const header = (
        <div>
          <p>This is the header</p>
          <Icon type="warning-solid" />
        </div>
      )
      const wrapper = mount(
        <Collapse {...collapseProps}>
          <Collapse.Panel {...collapsePanelProps} expandIconPosition={'left'} panelKey="panel_1" header={header}>
            折叠面板内容
          </Collapse.Panel>
        </Collapse>,
      )
      expect(wrapper.find(`.kd-collapse-panel-header`).at(0).childAt(0).childAt(1).childAt(0)).toHaveClassName(
        'kdicon-warning-solid',
      )
    })

    it('should render with correct panelKey', () => {
      const wrapper = mount(
        <Collapse {...collapseProps}>
          <Collapse.Panel header="header" panelKey="1" />
        </Collapse>,
      )
      expect(wrapper.find('Panel').prop('panelKey')).toEqual('1')
    })
  })

  // 7.component interaction(event)

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const Config = {
        compDefaultProps: {
          Collapse: {
            expandIconPosition: 'right',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={Config}>
          <Collapse>
            <Collapse.Panel header="This is the header1" panelKey="panel_1">
              折叠面板内容1
            </Collapse.Panel>
            <Collapse.Panel header="This is the header2" panelKey="panel_2">
              折叠面板内容2
            </Collapse.Panel>
          </Collapse>
        </ConfigProvider>,
      )

      expect(wrapper.find('.kd-collapse').first().find('.kd-collapse-panel-right')).toBeTruthy()
    })
  })

  // 9. ref test
  // describe('9. ref test', () => {
  //   it('should get Demo element from ref', () => {
  //     const ref = React.createRef()
  //     mount(
  //       <ConfigProvider ref={ref}>
  //         <Collapse>
  //           <Collapse.Panel header="This is the header1" panelKey="panel_1">
  //             折叠面板内容1
  //           </Collapse.Panel>
  //           <Collapse.Panel header="This is the header2" panelKey="panel_2">
  //             折叠面板内容2
  //           </Collapse.Panel>
  //         </Collapse>
  //       </ConfigProvider>,
  //     )
  //     expect(ref.current instanceof HTMLElement).toBe(true)
  //   })
  // })
})
