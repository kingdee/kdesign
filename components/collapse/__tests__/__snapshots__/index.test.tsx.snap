// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image 2. render test renders correctly 1`] = `
<div
  class="kd-collapse"
>
  <div
    class="kd-collapse-panel"
  >
    <span
      class="kd-collapse-panel-top"
    >
      <span
        class="kd-collapse-panel-left"
      >
        <span
          class="kd-collapse-panel-icon kd-collapse-panel-animation-expand"
        >
          <i
            class="kdicon kdicon-arrow-right-solid"
          />
        </span>
        <span
          class="kd-collapse-panel-header"
        >
          This is the header
        </span>
      </span>
      <span
        class="kd-collapse-panel-middle"
      />
      <span
        class="kd-collapse-panel-right"
      />
    </span>
    <div
      class="kd-collapse-panel-children-hide"
    >
      <div
        class=""
      >
        折叠面板内容
      </div>
    </div>
  </div>
</div>
`;
